File: algorithms/dynamic-programming/egg-drop/c/egg_drop.c
#include "egg_drop.h"
#include <limits.h>

int egg_drop(const int* arr, int n) {
    int eggs = arr[0], floors = arr[1];
    int dp[100][1000];
    for (int e = 0; e <= eggs; e++)
        for (int f = 0; f <= floors; f++)
            dp[e][f] = 0;
    for (int f = 1; f <= floors; f++) dp[1][f] = f;
    for (int e = 2; e <= eggs; e++) {
        for (int f = 1; f <= floors; f++) {
            dp[e][f] = INT_MAX;
            for (int x = 1; x <= f; x++) {
                int a = dp[e-1][x-1], b = dp[e][f-x];
                int worst = 1 + (a > b ? a : b);
                if (worst < dp[e][f]) dp[e][f] = worst;
            }
        }
    }
    return dp[eggs][floors];
}

File: algorithms/dynamic-programming/egg-drop/kotlin/EggDrop.kt
fun eggDrop(arr: IntArray): Int {
    val eggs = arr[0]; val floors = arr[1]
    val dp = Array(eggs + 1) { IntArray(floors + 1) }
    for (f in 1..floors) dp[1][f] = f
    for (e in 2..eggs) {
        for (f in 1..floors) {
            dp[e][f] = Int.MAX_VALUE
            for (x in 1..f) {
                val worst = 1 + maxOf(dp[e - 1][x - 1], dp[e][f - x])
                dp[e][f] = minOf(dp[e][f], worst)
            }
        }
    }
    return dp[eggs][floors]
}
